lib/rubocop/cop/lint/memorization_with_parameters.rb

```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks potentially buggy usages of the ||= operators that
+      # store dynamic values in instance variables based on method parameters.
+      #
+      # @example
+      #
+      #   # bad
+      #   def foo(bar)
+      #     @foo ||= baz(bar)
+      #   end
+      #
+      #   # bad
+      #   def foo(bar)
+      #     @foo ||= "Test #{bar}"
+      #   end
+      #
+      #   # good
+      #   def foo(bar)
+      #     @foo ||= {}
+      #     @foo[bar] ||= baz(bar)
+      #   end
+      class MemorizationWithParameters < Cop
+        MSG = 'Use an instance variable such as a Hash to store method' \
+              ' returns called with parameter(s).'.freeze
+
+        def_node_matcher :shadowed_parameter?, <<-PATTERN
```

Comment:
this will match any number of arguments (instead of just one) then check the assignment node to see if it uses any of the passed in arguments

Reply:
This might be a better solution:

```
        def_node_search :memoization, '(or_asgn (ivasgn _) $_)'
        def_node_search :contains_argument?, '(lvar %1)'

        def on_def(node)
          memoization(node.body) do |assignment|
            node.arguments.each do |arg|
              next unless contains_argument?(assignment, arg.children.first)

              add_offense(node)
              return
            end
          end
        end
```

Allows for the following test case:

```
  context 'when method contains more than just memoization' do
    subject(:source) do
      <<-'RUBY'.strip_indent
        def foo(bar)
          puts "Called with #{bar}"
          @foo ||= baz(bar)
        end
      RUBY
    end

    it { expect(cop.offenses.size).to eq(1) }
    it { expect(cop.offenses.first.message).to eq error_message }
  end
```

Reply:
I committed that so I could replace the original code.